substrate/pkg/controller/substrate/cluster/config.go

```diff
@@ -0,0 +1,259 @@
+	if err := c.uploadDirectories(ctx, substrate); err != nil {
+		return reconcile.Result{}, fmt.Errorf("uploading to S3 bucket, %w", err)
+	}
+	return reconcile.Result{}, nil
+}
+
+func (c *Config) Delete(ctx context.Context, substrate *v1alpha1.Substrate) (reconcile.Result, error) {
+	// delete the s3 bucket
+	if err := s3manager.NewBatchDeleteWithClient(c.S3).Delete(ctx, s3manager.NewDeleteListIterator(
+		c.S3, &s3.ListObjectsInput{Bucket: discovery.Name(substrate)})); err != nil {
+		return reconcile.Result{}, fmt.Errorf("deleting objects from bucket %s, %w", aws.StringValue(discovery.Name(substrate)), err)
+	}
+	if _, err := c.S3.DeleteBucketWithContext(ctx, &s3.DeleteBucketInput{Bucket: discovery.Name(substrate)}); err != nil {
+		return reconcile.Result{}, fmt.Errorf("deleting S3, %w", err)
+	}
+	logging.FromContext(ctx).Infof("Deleted S3 bucket %s", aws.StringValue(discovery.Name(substrate)))
+	return reconcile.Result{}, nil
+}
+
+func (c *Config) generateCerts(cfg *kubeadm.InitConfiguration, substrate *v1alpha1.Substrate) error {
+	cfg.CertificatesDir = path.Join(clusterCertsBasePath, substrate.Name, certPKIPath)
+	certTree, err := certsphase.GetDefaultCertList().AsMap().CertTree()
+	if err != nil {
+		return err
+	}
+	if err := certTree.CreateTree(cfg); err != nil {
+		return fmt.Errorf("error creating cert tree, %w", err)
+	}
+	return nil
+}
+
+func (c *Config) generateStaticPodManifests(cfg *kubeadm.InitConfiguration, substrate *v1alpha1.Substrate) error {
+	manifestDir := path.Join(clusterCertsBasePath, substrate.Name, clusterManifestPath)
+	// etcd phase adds cfg.CertificatesDir to static pod yaml for pods to read the certs from
+	cfg.CertificatesDir = certPKIPath
+	if err := etcdphase.CreateLocalEtcdStaticPodManifestFile(
+		manifestDir, "", cfg.NodeRegistration.Name, &cfg.ClusterConfiguration, &cfg.LocalAPIEndpoint, false); err != nil {
+		return fmt.Errorf("error creating local etcd static pod manifest file %w", err)
+	}
+	for _, componentName := range []string{
+		kubeadmconstants.KubeAPIServer,
+		kubeadmconstants.KubeControllerManager,
+		kubeadmconstants.KubeScheduler} {
+		err := controlplane.CreateStaticPodFiles(path.Join(clusterCertsBasePath, substrate.Name, clusterManifestPath), "",
+			&cfg.ClusterConfiguration, &cfg.LocalAPIEndpoint, false, componentName)
+		if err != nil {
+			return fmt.Errorf("creating static pod file for %v, %w", componentName, err)
+		}
+	}
+	return nil
+}
+
+func (c *Config) ensureS3Bucket(ctx context.Context, substrate *v1alpha1.Substrate) error {
+	if _, err := c.S3.CreateBucket(&s3.CreateBucketInput{
+		Bucket: discovery.Name(substrate),
+		CreateBucketConfiguration: &s3.CreateBucketConfiguration{
+			LocationConstraint: c.S3.Config.Region,
+		},
+	}); err != nil && err.(awserr.Error).Code() != s3.ErrCodeBucketAlreadyOwnedByYou {
+		return fmt.Errorf("creating S3 bucket, %w", err)
+	}
+	logging.FromContext(ctx).Infof("Ensured s3 bucket %s", aws.StringValue(discovery.Name(substrate)))
```

Comment:
Can we differentiate created vs found? This log line should say created. Inside the error statement, you should log "Found". There are a ton of examples of this.